compiler/ast.nim

```diff
@@ -767,6 +767,7 @@ type
     typ*: PType
     info*: TLineInfo
     flags*: TNodeFlags
+    expandedFrom*: PNode
```

Comment:
No. Keeping the original AST after a template expansion causes a memory explosion in principle.

Reply:
Would storing the line info where the parser ended after parsing each PNode also cause too much memory usage? Alternatively I imagine the parser could run again using the TLineInfo of a Pnode and output an AST which could be rendered.

Reply:
```nim


template fac(n: int): int =
  when n == 2: 2
  else: n*fac(n-1)

echo fac(4)
```

--> No matter **how** you store it, **if** you store it, the memory consumption is exponential for `fac`.

Reply:
Either Nim's peakmem usage is lying or I must be missing the point as I do not see any memory difference in this example.

### For `fac(4)` :
- `26788 lines; 0.517s; 31.707MiB peakmem` Stock Nim #devel
- `26798 lines; 0.429s; 31.75MiB` This PR

### For `fac(40)`:
- `26798 lines; 0.429s; 31.75MiB` Stock Nim #devel
- `26798 lines; 0.442s; 31.762MiB peakmem` # This PR


This implementation only uses the top AST node of a template to store reference to where the code, it does not propagate downwards to every single child node. So I think the cases you'd see memory increase is with many nested macros, but even then it's going to mostly just be the dud 8 bytes on all the PNodes.